src/ripple/app/tx/impl/AMMCreate.cpp

```diff
@@ -0,0 +1,340 @@
+    sb.insert(sleAMMRoot);
+
+    // Calculate initial LPT balance.
+    auto const lpTokens = ammLPTokens(amount, amount2, lptIss);
+
+    // Create ltAMM
+    auto ammSle = std::make_shared<SLE>(ammKeylet);
+    ammSle->setFieldU16(sfTradingFee, ctx_.tx[sfTradingFee]);
+    ammSle->setAccountID(sfAccount, *ammAccount);
+    ammSle->setFieldAmount(sfLPTokenBalance, lpTokens);
+    auto const& [issue1, issue2] = std::minmax(amount.issue(), amount2.issue());
+    ammSle->setFieldIssue(sfAsset, STIssue{sfAsset, issue1});
+    ammSle->setFieldIssue(sfAsset2, STIssue{sfAsset2, issue2});
+    // AMM creator gets the voting slot.
+    STArray voteSlots;
+    STObject voteEntry{sfVoteEntry};
+    voteEntry.setFieldU16(sfTradingFee, ctx_.tx[sfTradingFee]);
+    voteEntry.setFieldU32(sfVoteWeight, 100000);
+    voteEntry.setAccountID(sfAccount, account_);
+    voteSlots.push_back(voteEntry);
+    ammSle->setFieldArray(sfVoteSlots, voteSlots);
+    // AMM creator gets the auction slot for free.
+    auto& auctionSlot = ammSle->peekFieldObject(sfAuctionSlot);
+    auctionSlot.setAccountID(sfAccount, account_);
+    // current + sec in 24h
+    auto const expiration =
+        std::chrono::duration_cast<std::chrono::seconds>(
+            ctx_.view().info().parentCloseTime.time_since_epoch())
+            .count() +
+        TOTAL_TIME_SLOT_SECS;
+    auctionSlot.setFieldU32(sfExpiration, expiration);
+    auctionSlot.setFieldU32(sfDiscountedFee, 0);
+    auctionSlot.setFieldAmount(sfPrice, STAmount{lpTokens.issue(), 0});
+    sb.insert(ammSle);
+
+    // Add owner directory to link the root account and AMM object.
+    if (auto const page = sb.dirInsert(
+            keylet::ownerDir(*ammAccount),
+            ammSle->key(),
+            describeOwnerDir(*ammAccount));
+        !page)
+    {
+        JLOG(j_.debug()) << "AMM Instance: failed to insert owner dir";
+        return {tecDIR_FULL, false};
+    }
+
+    // Send LPT to LP.
+    auto res = accountSend(sb, *ammAccount, account_, lpTokens, ctx_.journal);
+    if (res != tesSUCCESS)
+    {
+        JLOG(j_.debug()) << "AMM Instance: failed to send LPT " << lpTokens;
+        return {res, false};
+    }
+
+    auto sendAndTrustSet = [&](STAmount const& amount) -> TER {
+        if (auto const res =
+                accountSend(sb, account_, *ammAccount, amount, ctx_.journal))
+            return res;
+        // Set AMM flag on AMM trustline
+        if (!isXRP(amount))
+        {
+            if (SLE::pointer sleRippleState =
+                    sb.peek(keylet::line(*ammAccount, amount.issue()));
+                !sleRippleState)
+                return tecINTERNAL;
+            else
+            {
+                auto const flags = sleRippleState->getFlags();
+                sleRippleState->setFieldU32(sfFlags, flags | lsfAMMNode);
+                sb.update(sleRippleState);
+            }
+        }
+        return tesSUCCESS;
+    };
+
+    // Send asset1.
+    res = sendAndTrustSet(amount);
```

Comment:
If we successfully sent the LPT to the LP but `sendAndTrustSet` fails, will the LP keep the LPT?
Is it possible to unroll the above step?

Reply:
The changes are committed to the `SandBox` and are only applied in `AMMCreate::doApply()` on success. None of the changes takes effect on failure.